cilium/cmd/helpers.go

```diff
@@ -366,3 +367,14 @@ func dumpConfig(Opts map[string]string) {
 		}
 	}
 }
+
+// initSupportedMapTypes triggers the BPF feature checks and gets information
+// about supported BPF map types. Not calling that function in BPF CT related
+// commands will lead to nil pointer errors.
+func initSupportedMapTypes() {
```

Comment:
Have you considered creating an instance of `ProbeManager` inside `pkg/datapath/linux/probes`, and then calling it directly from `pkg/bpf`? That would allow us to get rid of `initSupportedMapTypes()` and similar constructs.

Reply:
I have a concern about this suggestion: would this approach mean that as soon as any application or test imports `pkg/bpf`, it will trigger querying the kernel state? I wouldn't want to move us further away from the goal of being able to import and unit-test packages that rely upon `pkg/bpf` on non-linux developer environments, non-root, etc.

Reply:
That's a good point. My biggest concern is that in some code path we might forget to initialize `supportedMapTypes` which would lead to a runtime panic.

Reply:
That's a great idea @brb.

@joestringer I think that we can solve that problem by:

1. Moving `probes` package to `pkg/datapath/probe`.
2. Implementing the Linux kernel prober in the actual `probes_linux.go` file, which will build only on Linux.
3. Implementing some dummy "TODO" probers for the other platforms.

WDYT?

Reply:
I have a problem, folks.

I've tried having a global instance of `ProbeManager` and initializing it in `init` in maaany ways. Every time I had some issues. Apparently, the `pkg/datapath/linux/probes` package ended up being imported even when launching commands like `cilium `. And I had troubles with getting tests working. There are just too many cases where some commands/programs which are not cilium-agent end up importing that package.

So I think that we have two choices:
- I can try to analyze why that package gets imported when it really shouldn't and prevent it from being imported. But I think that solution would mean refactoring currently existing code dealing with maps. And it will mean postponing that PR for a longer time.
- Get back to having `initSupportedMapTyped` and calling it from cilium-agent and BPF-related CLI commands.

For sake of having a clean code, I'm more in favor for the first option.

Reply:
Is the former option tractable? I was under the impression that as soon as you import a package with an `init()` ("said package") into your program, the compiler will add the execution of that `init()` function into the program startup prior to `main()`, regardless of whether an individual runtime execution for the binary would execute code in said package or not. I'm no Go expert though so please feel free to prove me wrong in this thinking.

Honestly `init()` functions just make me nervous in general. I'm sure there's limited appropriate usage for simple initialization but this kind of case (including process execution & kernel probing and hence plenty of failure conditions) doesn't seem appropriate for it.

Reply:
Agreed on the `init()` function being inappropriate in this case.

Reply:
What should I do then?

- Have the global `ProbeManager` variable in `pkg/datapath/linux/probes`, call `ProbeManager` in `bpf/maps` initialize it in Daemon and CLI commands, call it in `pkg/bpf/map.go`?
- Have the global `ProbeManager` variable in `pkg/datapath/linux/probes`, initialize it in Daemon and CLI commands, fill up the `supportedMapTypes` variables in `pkg/bpf/map.go` during initialization of `ProbeManager`?
- Anything else?

Reply:
Broadly I think we have three options:

1. Initialize/probe via explicit call from caller of probes package which initializes static package variable inside probes package. Failure to do so may cause segfault. Explicit call returns an error type, so error handling for fatal probe error (any failure condition we can't handle, including process kill of bpftool etc.) is clearly handled if users properly use the library.
1. Initialize/probe via `sync.Once` as part of any exported functions from the probes package. Initializes the static package variable only at runtime for users of the function. Two options for handling fatal probe error:
   1. Every exported function may return an error, callers must handle these.
   1. Exported functions return default values if initialize failed. Not great as it's difficult to propagate the fatal probe error.
1. Initialize/probe via explicit ProbesManager object creation. All callers that wish to perform operations that require probing must, either directly or indirectly, provide this object. This is really only different from (1) or (2) specifically for the probes package. Callers of the package all the way up the stack must all eventually do some variation on (1) or (2): Initialize at some point explicitly (knowing that failure to do so may cause segfault), or initialize and cache the object from users.

This is the way I think about the comparison between options:

| Solution | Initialization cost | Error handling | Consequence for not initializing | Ability to test |
| --- | --- | --- | --- | --- |
| 1 | Once | Once | Potential crashes | Medium |
| 2 | Per-function | Per-function | None | Low |
| 3 | Flexible | Flexible | Varies | High |

(1) is better if you have a large number of operations and are not worried about library misuse.  Testing without the actual functionality requires creating mock initialization functions but it's a bit ugly.
(2) works pretty well if the number of operations is low, although the use of static variables does decrease the ability to unit-test, because the static variable is always configured. Testing can be done in a similar manner to (1), but if a test forgets to perform the fake initialization then it's more painful here as it will pull in the real implementation by default.
(3) allows balancing the +/- of the above, depending at which layer you perform the initialization. Recently in Cilium we have pushed towards initialization and explicit object creation in the daemon to coordinate all interaction. This makes it easier to mock out certain aspects of cilium operation for easier testing, but it comes at the cost of additional work and additional code from contributors. For CLI this pushes the responsibility into the CLI initialization code.

For this specific case, I look at the probeManager interface:

```
type probeManager interface {
	SystemConfigProbes() error
	GetSupportedMapTypes() *SupportedMapTypes
}
```

`SystemConfigProbes()` is an explicit call from the daemon once at the start for sanity-checking purposes, and `GetSupportedMapTypes()` is called from `Map.openOrCreate()`, `Map.Open()`, and `Map.CheckAndUpgrade()`. Map open or upgrade calls should not be highly frequent, so the cost of checking a synchronization primitive is not overbearing. Furthermore, most of these functions already return error codes so if the feature probing returns an error, it is easy to pass up (exception: `CheckAndUpgrade()` which might need a little more thinking).

Although longer term I hope that we can move more towards (3) for better testing, for this particular PR I'm not fussed between (1) or (2). I think that the risk of forgetting to initialize is fairly low in CLI if we share initialization for all BPF subprograms, and even if we do hit a segfault there it's easily found and has low impact on production users so (1) is fine by me. (2) involves a bit more error checking/handling but seems reasonable given the number of times we're likely to call these functions.

One variation of (3) would be using the (2) method inside `pkg/bpf` for the open/create/upgrade. Then, at runtime it'd only be the CLI tools that interact with BPF maps that execute the probe and you may be able to avoid the explicit initialization. WDYT?

Reply:
Thanks for the detailed reply! I didn't think about method (2) and using `sync.Once`.

I like the idea of mixing methods (3) and (2) - creating the `ProbeManager` object explicitly in `pkg/datapath/linux/requirements.go` and `pkg/bpf/map.go`, where in `map.go` some function would initialize and use `ProbeManager` and initialize supported map types variable, being finally called with `sync.Once.Do` on open/create/upgrade.

`CheckAndUpgrade` doesn't return error, but it's used only in one place outside tests `pkg/maps/ctmap/ctmap.go`, so I think that changing its signature is not that big problem.

I will push an update tomorrow or this week, thanks again!